### PR TITLE
Use soft references in SQL query engine to avoid OOM

### PR DESCRIPTION
**Pull request by luigidellaquila**

### CODE REVIEW

File: core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java [hunk 1]
```diff
@@ -575,6 +575,13 @@ public void change(final Object iCurrentValue, final Object iNewValue) {
       Integer.class, 500),
 
   // QUERY
+  QUERY_USE_SOFT_REFENCES_IN_RESULT_SET("query.useSofReferencesInResultSet",
```
Comment (luigidellaquila) [inline core/src/main/java/com/orientechnologies/orient/core/config/OGlobalConfiguration.java]: There is a typo here!!! `useSofReferences` should be `useSoftReferences`

**Comment by tglman** [comment]
Hi,

I'm not yet 100% convinced by this solution, for have this solution working i saw that we have to use `Xms == Xmx` that may work when we control the setup(server dist) but it won't work when we don't control the setup(embedding), as well for the case of embedding OOM may come from a different source so soft reference will solve only half of the problem, and probably introducing some not deterministic behaviour(really hard to reproduce in case of error), like a result set disappearing while  iterating it, due to some other part of an application being overloaded(by design or mistake).

Regards

**Comment by laa** [comment]
@tglman you are right that is in case of xms=xmx because our commercial server users suffer from this problem and we have several issues in a critical state where they ping us to solve it. This solution will solve all those issues.

**Comment by laa** [comment]
Seems like we resolved all questions with @tglman .

**Comment by tglman** [comment]
Hi,
From a quick private chat: if Xms != Xml the soft reference check is turned off, and if is on and is triggered a free of a soft reference the query get killed without not deterministic behaviour, this is good enough for me now.

File: core/src/main/java/com/orientechnologies/orient/core/engine/OMemoryAndLocalPaginatedEnginesInitializer.java [hunk 2]
```diff
@@ -73,6 +84,75 @@ private void configureDefaults() {
...
+        OGlobalConfiguration.QUERY_USE_SOFT_REFENCES_IN_RESULT_SET.setValue(false);
+      }
+    }
+  }
+
+  private long extractMemoryLimitInBytes(String limit) {
+    final Pattern pattern = Pattern.compile("((-Xms)|(-Xmx))(\\d+)([gGmMkK]?)");
+    final Matcher matcher = pattern.matcher(limit);
+    if (!matcher.find()) {
+      throw new IllegalArgumentException("Invalid value of memory limit was provided '" + limit + "'");
```
Comment (luigidellaquila) [inline core/src/main/java/com/orientechnologies/orient/core/engine/OMemoryAndLocalPaginatedEnginesInitializer.java]: why...? There is really no need to throw an exception if you don't explicitly define the memory limit...

File: core/src/main/java/com/orientechnologies/orient/core/engine/OMemoryAndLocalPaginatedEnginesInitializer.java [hunk 3]
```diff
@@ -73,6 +84,75 @@ private void configureDefaults() {
...
+    dimension = dimension.toLowerCase(Locale.ENGLISH);
+
+    if (dimension.equals("g")) {
+      bytes = bytes * 1024 * 1024 * 1024;
+    } else if (dimension.equals("m")) {
+      bytes = bytes * 1024 * 1024;
+    } else if (dimension.equals("k")) {
+      bytes = bytes * 1024;
+    } else {
+      throw new IllegalArgumentException("Invalid dimension of memory limit + '" + dimension + "'");
```
Comment (luigidellaquila) [inline core/src/main/java/com/orientechnologies/orient/core/engine/OMemoryAndLocalPaginatedEnginesInitializer.java]: same here, no need to throw an exception

File: core/src/main/java/com/orientechnologies/orient/core/engine/OMemoryAndLocalPaginatedEnginesInitializer.java (hunk 2 above)
Comment (luigidellaquila) [reply core/src/main/java/com/orientechnologies/orient/core/engine/OMemoryAndLocalPaginatedEnginesInitializer.java]: ah, ok, it's checked above...

File: core/src/main/java/com/orientechnologies/orient/core/sql/OSoftQueryResultTimSort.java [hunk 4]
```diff
@@ -0,0 +1,911 @@
...
+          newSize = Math.min(newSize, aLength >>> 1);
+
+        //noinspection unchecked
+        tmp = new SoftReference[newSize];
+        tmpLength = newSize;
+      }
+      return tmp;
+    }
+  }
+}
```
Comment (luigidellaquila) [inline core/src/main/java/com/orientechnologies/orient/core/sql/OSoftQueryResultTimSort.java]: Trusting you here...

**Pull request merged**